lib/view_model/active_record/update_data.rb

```diff
@@ -55,14 +160,158 @@ def initialize(values)
     end
 
     class Replace < self
-      def update_datas
+      def update_datas(referenced_updates)
         values
       end
     end
 
     class Functional < self
-      def update_datas
-        values.flat_map(&:values)
+      def update_datas(referenced_updates)
+        values.reject { |action| action.is_a?(FunctionalUpdate::Remove) }.flat_map(&:values)
+      end
+    end
+
+    class Parser < UpdateParser
+      attr_reader :parse_association
+
+      def initialize(association_data, blame_reference, parse_association:)
+        super(association_data, blame_reference)
+        @parse_association = parse_association
+      end
+
+      def action_schema
+        UpdateData::Schemas::COLLECTION_UPDATE
+      end
+
+      def parse_update_values(values)
+        values.map(&parse_association)
+      end
+
+      def parse_update_action(action)
+        ViewModel::Schemas.verify_schema!(UpdateData::Schemas::UPDATE_ACTION, action)
```

Comment:
Just a general comment on this set of types: Was never meant to be a grand abstract class hierarchy. The public interface is two "unrelated" parsers that handle collections, and their internals is the existing algorithm just parameterised. Though you're correct that it could be tidier, and I've put more effort in.